Mathlib/Algebra/Order/Incidence.lean

```diff
@@ -0,0 +1,645 @@
+/-
+Copyright (c) 2022 Alex J. Best, Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex J. Best, Yaël Dillies
+-/
+import Mathlib.Algebra.Algebra.Defs
+import Mathlib.Algebra.BigOperators.Intervals
+import Mathlib.Algebra.BigOperators.Ring
+import Mathlib.Algebra.Module.BigOperators
+import Mathlib.Algebra.Module.Pi
+import Mathlib.Order.Interval.Finset.Defs
+
+/-!
+# Incidence algebras
+
+Given a locally finite order `α` the incidence algebra over `α` is the type of functions from
+non-empty intervals of `α` to some algebraic codomain.
+This algebra has a natural multiplication operation whereby the product of two such functions
+is defined on an interval by summing over all divisions into two subintervals the product of the
+values of the original pair of functions.
+This structure allows us to interpret many natural invariants of the intervals (such as their
+cardinality) as elements of the incidence algebra. For instance the cardinality function, viewed as
+an element of the incidence algebra, is simply the square of the function that takes constant value
+one on all intervals. This constant function is called the zeta function, after
+its connection with the Riemann zeta function.
+The incidence algebra is a good setting for proving many inclusion-exclusion type principles, these
+go under the name Möbius inversion, and are essentially due to the fact that the zeta function has
+a multiplicative inverse in the incidence algebra, an inductively definable function called the
+Möbius function that generalizes the Möbius function in number theory.
```

Comment:
There's a bunch of definitions in this file, I'd say a `## Main definitions` would be very valuable